src/DynamoCoreWpf/ViewModels/Menu/TrustedPathViewModel.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using Dynamo.Core;
+using Dynamo.Interfaces;
+using Dynamo.Wpf.Properties;
+using DelegateCommand = Dynamo.UI.Commands.DelegateCommand;
+using Dynamo.Models;
+using System.Windows.Data;
+using System.Globalization;
+using System.Linq;
+using Dynamo.Configuration;
+
+namespace Dynamo.ViewModels
+{
+
+    public class TrustedPathEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Indicate whether user wants to add the current path to the list
+        /// </summary>
+        public bool Cancel { get; set; }
+
+        /// <summary>
+        /// Indicate the path to the custom packages folder
+        /// </summary>
+        public string Path { get; set; }
+    }
+
+    public class TrustedPathViewModel : ViewModelBase
+    {
+        public ObservableCollection<string> TrustedLocations { get; private set; }
+        [Obsolete("SelectedIndex is no longer referenced, do not use.")]
+        public int SelectedIndex { get; set; }
+       
+        public event EventHandler<TrustedPathEventArgs> RequestShowFileDialog;
+        public virtual void OnRequestShowFileDialog(object sender, TrustedPathEventArgs e)
+        {
+            if (RequestShowFileDialog != null)
+            {
+                RequestShowFileDialog(sender, e);
+            }
+        }
+
+        public DelegateCommand AddPathCommand { get; private set; }
+        public DelegateCommand DeletePathCommand { get; private set; }
+        public DelegateCommand UpdatePathCommand { get; private set; }
+        public DelegateCommand SaveSettingCommand { get; private set; }
+
+        public TrustedPathViewModel()
+        { 
+            InitializeTrustedLocations();
+            InitializeCommands();
+        }
+        /// <summary>
+        /// This constructor overload has been added for backwards comptability.
+        /// </summary>
+        /// <param name="setting"></param>
+        public TrustedPathViewModel(IPreferences setting)
+        {
+            InitializeTrustedLocations();
+            InitializeCommands();
+        }
+        private void InitializeCommands() 
+        {
+            AddPathCommand = new DelegateCommand(p => InsertPath());
+            DeletePathCommand = new DelegateCommand(p => RemovePathAt(ConvertPathToIndex(p)), p => CanDelete(ConvertPathToIndex(p)));
+            UpdatePathCommand = new DelegateCommand(p => UpdatePathAt(ConvertPathToIndex(p)), p => CanUpdate(ConvertPathToIndex(p)));
+            SaveSettingCommand = new DelegateCommand(CommitChanges);
+        }
+
+        private int ConvertPathToIndex(object path)
+        {
+            if(path is int pint)
+            {
+                return pint;
+            }
+            return TrustedLocations.IndexOf(path as string);
```

Comment:
Do we really need to handle both strings and integers when managing CRUD operations ?
I would suggest we only handle strings, basically only keep line 78